planning/MartenV4.md

```diff
@@ -0,0 +1,465 @@
+# Marten v4 Ideas
+
+## Pulling it Off
+
+We've got the typical problems of needing to address incoming pull requests and bug issues in master while probably needing to have a long lived branch for *v4*. 
+
+As an initial plan, let's:
+
+1. Start with the unit testing improvements as a way to speed up the build before we dive into much more of this? *This is in progress with about a 25% reduction in test throughput time so far [in this pull request](https://github.com/JasperFx/marten/pull/1463)*
+1. Do a bug sweep v3.12 release to address as many of the tactical problems as we can before branching to v4
+1. ~~Possibly do a series of v4, then v5 releases to do this in smaller chunks? We've mostly said do the event store as v4, then Linq improvements as v5~~ -- *Nope, full speed ahead with a large v4 release in order to do as many breaking changes as possible in one release*
+1. [Extract the generic database manipulation code to its own library](https://github.com/JasperFx/marten/issues/1467) to clean up Marten, and speed up our builds to make the development work be more efficient.
+1. Do the Event Store v4 work in a separate project built as an add on from the very beginning, but leave the existing event store in place. That would enable us to do a lot of work *and* mostly be able to keep that work in master so we don't have long-lived branch problems. Break open the event store improvement work because that's where most of the interest is for this release.
+
+
+## Miscellaneous Ideas
+
+* Look at some kind of object pooling for the `DocumentSession` / `QuerySession` objects?
+* Ditch the document by document type schema configuration where Document *A* can be in one schema, and Document "B" is in another schema. Do that, and I think we open the door for multi-tenancy by schema. 
+* ~~Eliminate `ManagedConnection` altogether. I think it results in unnecessary object allocations and it's causing more harm that help as it's been extended over time.~~ **After studying that more today, it's just too damn embedded. At least try to kill off the `Execute` methods that take in a Lambda**. See [this GitHub issue](https://github.com/JasperFx/marten/issues/1469).
+* ~~Can we consider ditching < .Net Core or .Net v5 for v4?~~ **The probable answer is "no," so let's just take this off the table.**
```

Comment:
Yes, I think that knowing the last releases of Microsoft is risky to assume that it will just work and all people will move to v5 immediately.